File: paleo_workbench/ui/pages/map_edit_items.py
from __future__ import annotations

from typing import Any

from PySide6.QtCore import QPointF, QRectF, Qt
from PySide6.QtGui import QBrush, QColor, QPen, QPolygonF
from PySide6.QtWidgets import (
    QGraphicsEllipseItem,
    QGraphicsPolygonItem,
    QGraphicsRectItem,
)

from paleo_workbench.ui import tokens

# Default well marker radius in scene/map units.
WELL_RADIUS = 0.4
# Vertex handle half-size in scene/map units.
VERTEX_HANDLE_HALF = 0.35

_FACIES_FILL = QColor(tokens.PRIMARY)
_FACIES_FILL.setAlpha(70)
_FACIES_PEN = QPen(QColor(tokens.PRIMARY), 0)  # cosmetic width
_WELL_FILL = QBrush(QColor(tokens.TEAL))
_WELL_PEN = QPen(QColor(tokens.TEXT_DARK), 0)
_HANDLE_FILL = QBrush(QColor(tokens.TEAL))
_HANDLE_PEN = QPen(QColor(tokens.TEXT_DARK), 0)


class FeatureItemMixin:
    """Shared identity / export surface for map feature graphics items."""

    feature_id: str
    kind: str

    def to_record(self) -> dict[str, Any]:
        raise NotImplementedError


class VertexHandleItem(QGraphicsRectItem):
    """Draggable handle for a single ring vertex (scene-managed, not a feature)."""

    def __init__(
        self,
        feature_id: str,
        vertex_index: int,
        x: float,
        y: float,
        half: float = VERTEX_HANDLE_HALF,
        parent=None,
    ):
        h = float(half)
        super().__init__(QRectF(-h, -h, 2 * h, 2 * h), parent)
        self.feature_id = str(feature_id)
        self.vertex_index = int(vertex_index)
        self.setPos(float(x), float(y))
        self.setBrush(_HANDLE_FILL)
        self.setPen(_HANDLE_PEN)
        self.setZValue(100)
        self.setFlag(QGraphicsRectItem.GraphicsItemFlag.ItemIsSelectable, True)
        self.setFlag(QGraphicsRectItem.GraphicsItemFlag.ItemIsMovable, False)
        self.setAcceptedMouseButtons(Qt.MouseButton.LeftButton)


class FaciesPolygonItem(QGraphicsPolygonItem, FeatureItemMixin):
    """Facies polygon feature on the edit scene."""

    def __init__(
        self,
        feature_id: str,
        coordinates: list[list[float]],
        name: str = "",
        style: dict[str, Any] | None = None,
        parent=None,
    ):
        polygon = QPolygonF([QPointF(float(p[0]), float(p[1])) for p in coordinates])
        super().__init__(polygon, parent)
        self.feature_id = feature_id
        self.kind = "facies"
        self._name = name or ""
        self._style = dict(style or {})
        self._coordinates = [[float(p[0]), float(p[1])] for p in coordinates]
        self.setBrush(QBrush(_FACIES_FILL))
        self.setPen(_FACIES_PEN)
        self.setZValue(10)
        self.setFlag(QGraphicsPolygonItem.GraphicsItemFlag.ItemIsSelectable, True)

    def coordinates(self) -> list[list[float]]:
        return [list(p) for p in self._coordinates]

    def set_coordinates(self, coordinates: list[list[float]]) -> None:
        """Replace ring coordinates and refresh the polygon path."""
        self._coordinates = [[float(p[0]), float(p[1])] for p in coordinates]
        self.setPolygon(QPolygonF([QPointF(p[0], p[1]) for p in self._coordinates]))
        self.setPos(0.0, 0.0)

    def translate_by(self, dx: float, dy: float) -> None:
        """Shift polygon vertices and update the graphics path."""
        dx_f = float(dx)
        dy_f = float(dy)
        for p in self._coordinates:
            p[0] += dx_f
            p[1] += dy_f
        self.setPolygon(QPolygonF([QPointF(p[0], p[1]) for p in self._coordinates]))
        self.setPos(0.0, 0.0)

    def to_record(self) -> dict[str, Any]:
        return {
            "id": self.feature_id,
            "kind": self.kind,
            "name": self._name,
            "coordinates": [list(p) for p in self._coordinates],
            "style": dict(self._style),
        }


class WellPointItem(QGraphicsEllipseItem, FeatureItemMixin):
    """Well point feature as a small ellipse centered on map coordinates."""

    def __init__(
        self,
        feature_id: str,
        x: float,
        y: float,
        name: str = "",
        radius: float = WELL_RADIUS,
        parent=None,
    ):
        r = float(radius)
        super().__init__(QRectF(float(x) - r, float(y) - r, 2 * r, 2 * r), parent)
        self.feature_id = feature_id
        self.kind = "well"
        self._name = name or ""
        self._x = float(x)
        self._y = float(y)
        self._radius = r
        self.setBrush(_WELL_FILL)
        self.setPen(_WELL_PEN)
        self.setZValue(20)
        self.setFlag(QGraphicsEllipseItem.GraphicsItemFlag.ItemIsSelectable, True)

    def translate_by(self, dx: float, dy: float) -> None:
        """Shift well point and update the ellipse rect."""
        self._x = float(self._x) + float(dx)
        self._y = float(self._y) + float(dy)
        r = self._radius
        self.setRect(QRectF(self._x - r, self._y - r, 2 * r, 2 * r))
        self.setPos(0.0, 0.0)

    def to_record(self) -> dict[str, Any]:
        return {
            "id": self.feature_id,
            "kind": self.kind,
            "name": self._name,
            "coordinates": [self._x, self._y],
        }

File: libs/ui_data_core/qt/src/map_edit_items.cpp
#include "pwb/ui_data_qt/map_edit_items.hpp"

#include <QBrush>
#include <QColor>
#include <QFont>
#include <QPen>

namespace pwb::ui_data_qt {
namespace {

// tokens.py literals — the Python module constructs these QColors once at
// import from the static (light) token constants, so they are not
// theme-bound. Same values here.
QBrush facies_fill() {
    QColor c(QStringLiteral("#0b5563"));  // PRIMARY, alpha 70
    c.setAlpha(70);
    return QBrush(c);
}
QPen facies_pen() {
    return QPen(QColor(QStringLiteral("#0b5563")), 0);  // cosmetic
}
QBrush teal_fill() {
    return QBrush(QColor(QStringLiteral("#0f766e")));  // TEAL
}
QPen dark_pen() {
    return QPen(QColor(QStringLiteral("#101820")), 0);  // TEXT_DARK
}
QPen line_pen() {
    QPen p(QColor(QStringLiteral("#a65313")), 0);  // ACCENT
    p.setCosmetic(true);
    p.setWidth(2);
    return p;
}
QColor label_color() { return QColor(QStringLiteral("#101820")); }

}  // namespace

VertexHandleItem::VertexHandleItem(std::string feature_id_, int vertex_index_,
                                   double x, double y, double half,
                                   QGraphicsItem* parent, int part_index_,
                                   int ring_index_)
    : QGraphicsRectItem(QRectF(-half, -half, 2 * half, 2 * half), parent),
      feature_id(std::move(feature_id_)),
      vertex_index(vertex_index_),
      part_index(part_index_),
      ring_index(ring_index_) {
    setPos(x, y);
    setBrush(teal_fill());
    setPen(dark_pen());
    setZValue(100);
    setFlag(GraphicsItemFlag::ItemIsSelectable, true);
    setFlag(GraphicsItemFlag::ItemIsMovable, false);
    setAcceptedMouseButtons(Qt::MouseButton::LeftButton);
}

FaciesPolygonItem::FaciesPolygonItem(
    std::string feature_id, const domain::Json& coordinates,
    std::string name, const domain::Json& style, const domain::Json& extras,
    std::string geometry_type, const domain::Json* geometry_coordinates,
    QGraphicsItem* parent)
    : QGraphicsPathItem(parent),
      model_(std::move(feature_id), coordinates, std::move(name), style,
             extras, std::move(geometry_type), geometry_coordinates) {
    refresh_path();
    setBrush(facies_fill());
    setPen(facies_pen());
    setZValue(10);
    setFlag(GraphicsItemFlag::ItemIsSelectable, true);
}

void FaciesPolygonItem::refresh_path() {
    QPainterPath path;
    path.setFillRule(Qt::FillRule::OddEvenFill);
    for (const auto& polygon : model_.polygons) {
        for (const auto& ring : polygon) {
            if (ring.size() < 3) continue;
            path.moveTo(ring[0][0], ring[0][1]);
            for (std::size_t i = 1; i < ring.size(); ++i)
                path.lineTo(ring[i][0], ring[i][1]);
            path.closeSubpath();
        }
    }
    setPath(path);
}

void FaciesPolygonItem::set_ring_coordinates(
    int part_index, int ring_index, const domain::Json& coordinates) {
    model_.set_ring_coordinates(part_index, ring_index, coordinates);
    refresh_path();
    setPos(0.0, 0.0);
}

void FaciesPolygonItem::set_coordinates(const domain::Json& coordinates) {
    model_.set_coordinates(coordinates);
    refresh_path();
    setPos(0.0, 0.0);
}

void FaciesPolygonItem::translate_by(double dx, double dy) {
    model_.translate_by(dx, dy);
    refresh_path();
    setPos(0.0, 0.0);
}

WellPointItem::WellPointItem(std::string feature_id, double x, double y,
                             std::string name, double radius,
                             QGraphicsItem* parent)
    : QGraphicsEllipseItem(QRectF(x - radius, y - radius, 2 * radius,
                                  2 * radius),
                           parent),
      model_(std::move(feature_id), x, y, std::move(name), radius) {
    setBrush(teal_fill());
    setPen(dark_pen());
    setZValue(20);
    setFlag(GraphicsItemFlag::ItemIsSelectable, true);
}

void WellPointItem::translate_by(double dx, double dy) {
    model_.translate_by(dx, dy);
    const double r = model_.radius;
    setRect(QRectF(model_.x - r, model_.y - r, 2 * r, 2 * r));
    setPos(0.0, 0.0);
}

LineItem::LineItem(std::string feature_id,
                   const pwb::ui_data_core::MapRing& coordinates,
                   std::string name, QGraphicsItem* parent)
    : QGraphicsPathItem(parent),
      model_(std::move(feature_id), coordinates, std::move(name)) {
    setPen(line_pen());
    setBrush(QBrush(Qt::BrushStyle::NoBrush));
    setZValue(15);
    setFlag(GraphicsItemFlag::ItemIsSelectable, true);
    rebuild_path();
}

void LineItem::rebuild_path() {
    QPainterPath path;
    const auto& pts = model_.points;
    if (!pts.empty()) {
        path.moveTo(pts[0][0], pts[0][1]);
        for (std::size_t i = 1; i < pts.size(); ++i)
            path.lineTo(pts[i][0], pts[i][1]);
    }
    setPath(path);
}

void LineItem::set_coordinates(const domain::Json& coordinates) {
    model_.set_coordinates(coordinates);
    rebuild_path();
    setPos(0.0, 0.0);
}

void LineItem::translate_by(double dx, double dy) {
    model_.translate_by(dx, dy);
    rebuild_path();
    setPos(0.0, 0.0);
}

LabelItem::LabelItem(std::string feature_id, double x, double y,
                     std::string text, std::string name,
                     QGraphicsItem* parent)
    : QGraphicsSimpleTextItem(
          QString::fromStdString(!text.empty() ? text : name), parent),
      model_(std::move(feature_id), x, y, std::move(text), std::move(name)) {
    setBrush(QBrush(label_color()));
    QFont font;
    font.setPointSize(10);
    setFont(font);
    setZValue(30);
    setFlag(GraphicsItemFlag::ItemIsSelectable, true);
    setPos(x, y);
}

void LabelItem::translate_by(double dx, double dy) {
    model_.translate_by(dx, dy);
    setPos(model_.x, model_.y);
}

void LabelItem::set_property(std::string_view key, const domain::Json& value) {
    model_.set_property(key, value);
    if (key == "text") setText(QString::fromStdString(model_.text));
}

}  // namespace pwb::ui_data_qt
